### PR TITLE
[ranking] adding functionality to gather global references for ranking

### PR DESCRIPTION
**Pull request by Numbers88s**
Current PoC for adding functionality to gather global references for ranking

## Test plan
Added some test and also manually tested.

### CODE REVIEW

File: migrations/frontend/1674600319_add_codeintel_global_ranks/up.sql [hunk 1]
```diff
@@ -0,0 +1,14 @@
+CREATE TABLE IF NOT EXISTS codeintel_global_ranks (
```
Comment (efritz) [inline migrations/frontend/1674600319_add_codeintel_global_ranks/up.sql:1]: Doesn't look like we're going to use this table?

File: .vscode/settings.json [hunk 2]
```diff
@@ -37,7 +37,7 @@
     "build.allowImplicitNetworkAccess": true,
     "local": "github.com/sourcegraph/sourcegraph"
   },
-  "jest.jestCommandLine": "pnpm run test",
+  "jest.jestCommandLine": "pnpm test",
```
Comment (efritz) [inline .vscode/settings.json:40]: wot this?

File: enterprise/internal/codeintel/uploads/ranking.go [hunk 3]
```diff
@@ -14,16 +15,161 @@ import (
...
+
+			start = stop + 1
+			stop = stop + 10000
+		}
+	}
+
+	countMap, err := redisStore.HGetAll("graph:globalfiles:counts").StringMap()
+	if err != nil {
+		return err
+	}
```
Comment (efritz) [inline enterprise/internal/codeintel/uploads/ranking.go:147]: We'll need to stream these into Postgres and do a final merge there instead of loading all of the counts into worker memory. We already do a similar 2-step process in the ranking service.

File: enterprise/internal/codeintel/uploads/ranking.go [hunk 4]
```diff
@@ -14,16 +15,161 @@ import (
...
+				redisCountMap[k] = v
+			}
+
+			err = redisStore.HMSet("graph:globalfiles:counts", redisCountMap).Err()
+			if err != nil {
+				return err
+			}
+
+			start = stop + 1
+			stop = stop + 10000
```
Comment (efritz) [inline enterprise/internal/codeintel/uploads/ranking.go:140]: Occurrences of these values should use package-level constants.

File: enterprise/internal/codeintel/uploads/ranking.go [hunk 5]
```diff
@@ -14,16 +15,161 @@ import (
...
+	err = redisStore.Del("graph:globalfiles:counts")
+	if err != nil {
+		return err
+	}
+
+	for k, v := range countMap {
+		fmt.Printf("OHBOY OHBOY OHBOY %s: %s \n", k, v)
+	}
+
+	redisStore.Incr("ranking:graph:processed")
```
Comment (efritz) [inline enterprise/internal/codeintel/uploads/ranking.go:158]: Should jump to the number of export at the start of reduction, not increase by one.

File: enterprise/internal/codeintel/uploads/internal/lsifstore/lsifstore_ranking.go [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+	otlog "github.com/opentracing/opentracing-go/log"
+	"github.com/sourcegraph/scip/bindings/go/scip"
+	"google.golang.org/protobuf/proto"
+
+	db "github.com/sourcegraph/sourcegraph/enterprise/internal/codeintel/uploads/internal/store"
+	"github.com/sourcegraph/sourcegraph/internal/database/basestore"
+	"github.com/sourcegraph/sourcegraph/internal/observation"
+)
+
+func (s *store) CreateDefinitionsAndReferencesForRanking(
```
Comment (efritz) [inline enterprise/internal/codeintel/uploads/internal/lsifstore/lsifstore_ranking.go:17]: @Numbers88s I think there is already a ScanDocuments function that could be re-used for this purpose. Can we consolidate (is it in the same package or am I mis-remembering)?

File: enterprise/internal/codeintel/uploads/ranking.go [hunk 7]
```diff
@@ -40,18 +186,33 @@ func (s *Service) SerializeRankingGraph(ctx context.Context, numRankingRoutines
...
+				// 		log.Int("id", upload.ID),
+				// 		log.String("repo", upload.Repo),
+				// 		log.String("root", upload.Root),
+				// 		log.Error(err),
+				// 	)
+
+				// 	return err
+				// }
+
+				if err := s.lsifstore.CreateDefinitionsAndReferencesForRanking(ctx, upload, s.setDefinitionsAndReferencesForUpload); err != nil {
```
Comment (efritz) [inline enterprise/internal/codeintel/uploads/ranking.go:201]: Can we replace the body of `serializeAndPersistRankingGraphForUpload` with this line so that this method stays (largely) unchanged in this PR?

File: enterprise/internal/codeintel/uploads/ranking.go [hunk 8]
```diff
@@ -14,16 +15,161 @@ import (
...
 )
 
-func (s *Service) SerializeRankingGraph(ctx context.Context, numRankingRoutines int) error {
-	if s.rankingBucket == nil {
+const (
+	rankingDefinitionHash = "graphDefs:dev"
+	rankingReferenceHash  = "graphRefs:dev"
+)
+
+func (s *Service) ReduceRankingGraph(ctx context.Context, numRankingRoutines int) error {
```
Comment (efritz) [inline enterprise/internal/codeintel/uploads/ranking.go:30]: This graph reducer is analogous to the `ranking/internal/background/loader.go` service and the logic here can replace the logic in that file. This will allow us to keep the old infrastructure in Postgres while updating it to deal with file reference count graphs rather than Spark output.

[These lines](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@3acd8f31db54a5f4069eefd0acc4eacbbd86d883/-/blob/enterprise/internal/codeintel/ranking/internal/background/loader.go?L71-115) define a goroutine that reads the file listing from GCS. [These lines](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@3acd8f31db54a5f4069eefd0acc4eacbbd86d883/-/blob/enterprise/internal/codeintel/ranking/internal/background/loader.go?L117-186) define gorotines that read each of the files and set some data into the database via `BulkSetDocumentRanks`.

This is the same method you should end up calling to serialize file reference counts (replacing the `HMSet` below).

File: enterprise/internal/codeintel/uploads/ranking.go [hunk 9]
```diff
@@ -14,16 +15,161 @@ import (
...
+	}
+
+	redisStore.Incr("ranking:graph:processed")
+
+	err = s.store.SetGlobalRanks(ctx, countMap)
+	if err != nil {
+		return err
+	}
+
+	return nil
```
Comment (efritz) [inline enterprise/internal/codeintel/uploads/ranking.go:165]: We'll want to do this async as well. See `ranking/internal/background/merger.go` for a routine that continually smushes two dictionaries of ranks together.

I think we can either use the methods defined there for most database interactions, or be able to use them with minor tweaks.

**Review by efritz — COMMENTED** [review]
@Numbers88s I left a few comments about vague direction to migrate redis -> postgres. These probably aren't understandable on their own and will better serve as little notes when we discus it synchronously.

Please read the `merger.go` and `loader.go` background routines as early as you can, and we can discuss some general strategy in our 1-1 on Monday and pair on it more intensely Tuesday.

The more familiar you are with what's currently there in the loader/merger the easier it will be to move to Postgres. As a simple overview:

- Spark outputs some junk into a GCS bucket
- (Loader) We consume each of those files into Postgres, splitting the file contents over several rows so that they are 
- (Merger) We consume each of these rows merge them into a single canonical row in another table.

What I'd like to see is a similar structure:

- When serializing a SCIP document, we serialize its definitions into a Postgres table (keyed by the upload so we can also delete it later when the upload goes away), as well as references to other definitions. This should be some number of rows like (target repo, hash of {target path -> reference count}). This is similar to the current serializer and loader processes.
- Concurrently, a background task (the loader) reads target repo/target path counts from the table written by the previous worker and aggregates them into the actual table. Since the aggregation here is the sum, we can directly update the codeintel path ranks table. This is similar to what the merger does today.

File: enterprise/internal/codeintel/uploads/ranking.go (hunk 4 above)
Comment (Numbers88s) [reply enterprise/internal/codeintel/uploads/ranking.go:140]: Yup. Done in wip branch.

File: lib/codeintel/precise/types.go [hunk 10]
```diff
@@ -254,3 +254,10 @@ type GroupedBundleDataMaps struct {
 	Packages          []Package
 	PackageReferences []PackageReference
 }
+
+type SymbolLocation struct {
+	// The symbol name which is unique within the document.
+	SymbolID string
+	// The path to the document containing the symbol.
+	Path string
+}
```
Comment (efritz) [inline lib/codeintel/precise/types.go:263]: Now unused?

File: migrations/frontend/1676319972_add_codeintel_rank_defintions_and_references/up.sql [hunk 11]
```diff
@@ -0,0 +1,15 @@
+CREATE TABLE codeintel_ranking_definitions (
+    upload_id int NOT NULL,
+    symbol_name text NOT NULL,
+    repository text NOT NULL,
+    document_root text NOT NULL,
+    document_path text NOT NULL
```
Comment (efritz) [inline migrations/frontend/1676319972_add_codeintel_rank_defintions_and_references/up.sql:6]: At this point you should only have path - path should be prepended here as there's no reason to keep them distinct.

File: migrations/frontend/1676319972_add_codeintel_rank_defintions_and_references/up.sql [hunk 12]
```diff
@@ -0,0 +1,15 @@
...
+);
+
+CREATE INDEX IF NOT EXISTS codeintel_ranking_definitions_upload_id ON codeintel_ranking_definitions (upload_id);
+CREATE INDEX IF NOT EXISTS codeintel_ranking_definitions_symbol_name ON codeintel_ranking_definitions (symbol_name);
+
+CREATE TABLE codeintel_ranking_references (
+    upload_id int NOT NULL,
```
Comment (efritz) [inline migrations/frontend/1676319972_add_codeintel_rank_defintions_and_references/up.sql:13]: Will want an index here. Will also want a graph key and a processed flag.

File: migrations/frontend/1676399623_add_codeintel_rank_path_counts_inputs/up.sql [hunk 13]
```diff
@@ -0,0 +1,9 @@
+CREATE TABLE IF NOT EXISTS codeintel_ranking_path_counts_inputs (
+    repository text NOT NULL,
+    document_root text NOT NULL,
+    document_path text NOT NULL,
+    count int NOT NULL,
+    graph_key text NOT NULL,
+    processed boolean NOT NULL DEFAULT false
+);
```
Comment (efritz) [inline migrations/frontend/1676399623_add_codeintel_rank_path_counts_inputs/up.sql:8]: Once we've gotten the queries down, I'll leave it as an exercise to to the reader to select indices :)

File: enterprise/internal/codeintel/uploads/internal/background/job_graph_exporter.go [hunk 14]
```diff
@@ -30,3 +30,39 @@ func NewRankingGraphExporter(
...
+
+func NewRankingGraphMapper(
+	observationCtx *observation.Context,
+	uploadsService UploadService,
+	numRankingRoutines int,
+	interval time.Duration,
+) goroutine.BackgroundRoutine {
+	return goroutine.NewPeriodicGoroutine(
+		context.Background(),
+		"pagerank.graph-mapper", "maps graph",
```
Comment (efritz) [inline enterprise/internal/codeintel/uploads/internal/background/job_graph_exporter.go:42]: mmm yes descriptive :)

File: enterprise/internal/codeintel/uploads/internal/store/store.go [hunk 15]
```diff
@@ -144,3 +154,131 @@ func (s *store) transact(ctx context.Context) (*store, error) {
 func (s *store) Done(err error) error {
 	return s.db.Done(err)
 }
+
+const batchNumber = 10000
+
+// TODO: Move it out of here and the ranking service
+func (s *store) SetGlobalRanks(ctx context.Context, ranks map[string]string) (err error) {
```
Comment (efritz) [inline enterprise/internal/codeintel/uploads/internal/store/store.go:161]: This appears unused now.

File: enterprise/internal/codeintel/uploads/internal/lsifstore/lsifstore_ranking.go [hunk 16]
```diff
@@ -0,0 +1,375 @@
...
+
+	return nil
+}
+
+const insertPathCountInputsQuery = `
+WITH refs AS (
+	SELECT
+		unnest(symbol_names) AS symbol_names
+	FROM codeintel_ranking_references
+	WHERE upload_id = %s
```
Comment (efritz) [inline enterprise/internal/codeintel/uploads/internal/lsifstore/lsifstore_ranking.go:168]: As we talked about, you can drop the `WHERE` condition here and simply select a bunch of ranking references to process. See [this query](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@48f5a82b5a419f56fb4a6093c74ec916d989ce1c/-/blob/enterprise/internal/codeintel/ranking/internal/store/store.go?L252:7&subtree=true) for a sample.

File: enterprise/internal/codeintel/uploads/ranking.go [hunk 17]
```diff
@@ -14,15 +15,101 @@ import (
...
+	}
+
+	for _, uid := range referencesByUploadID {
+		// for {
+		uploadID, err := strconv.Atoi(uid)
+		if err != nil {
+			return err
+		}
+
+		if err := s.lsifstore.InsertPathCountInputs(ctx, uploadID); err != nil {
```
Comment (efritz) [inline enterprise/internal/codeintel/uploads/ranking.go:89]: The entire enclosing method can become a call to this function that takes a batch size, after the modifications described [here](https://github.com/sourcegraph/sourcegraph/pull/47233/files?diff=split&w=1#r1106560021).

File: enterprise/internal/codeintel/uploads/internal/lsifstore/lsifstore_ranking.go [hunk 18]
```diff
@@ -0,0 +1,375 @@
...
+	"google.golang.org/protobuf/proto"
+
+	db "github.com/sourcegraph/sourcegraph/enterprise/internal/codeintel/uploads/internal/store"
+	"github.com/sourcegraph/sourcegraph/enterprise/internal/codeintel/uploads/shared"
+	"github.com/sourcegraph/sourcegraph/internal/database/basestore"
+	"github.com/sourcegraph/sourcegraph/internal/database/batch"
+	"github.com/sourcegraph/sourcegraph/internal/observation"
+)
+
+func (s *store) InsertDefinitionsAndReferencesForRanking(
```
Comment (efritz) [inline enterprise/internal/codeintel/uploads/internal/lsifstore/lsifstore_ranking.go:21]: This should be called something more generic, like `forEachDocument`

File: enterprise/internal/codeintel/uploads/internal/lsifstore/lsifstore_ranking.go [hunk 19]
```diff
@@ -0,0 +1,375 @@
...
+
+func (s *store) InsertPathRanks(ctx context.Context, graphKey string, batchSize int) (err error) {
+	ctx, _, endObservation := s.operations.insertPathRanks.With(ctx, &err, observation.Args{LogFields: []otlog.Field{
+		otlog.String("graphKey", graphKey),
+	}})
+	defer endObservation(1, observation.Args{})
+
+	if err := s.db.Exec(ctx, sqlf.Sprintf(`TRUNCATE TABLE codeintel_path_ranks`)); err != nil {
+		return err
+	}
```
Comment (efritz) [inline enterprise/internal/codeintel/uploads/internal/lsifstore/lsifstore_ranking.go:190]: We'll want to drop this and modify the query below to work similarly to how the merger query works. This query might get flipped around a bit (we can use `all_current_ranks` only if there's a conflict we need to merge) to simplify the intent. We can chat about this part sync.

File: enterprise/internal/codeintel/uploads/internal/lsifstore/lsifstore_ranking.go [hunk 20]
```diff
@@ -0,0 +1,375 @@
...
+	upload_id,
+	symbol_name
+	repository,
+	document_root,
+	document_path
+FROM codeintel_ranking_definitions
+WHERE symbol_name = ANY(%s)
+`
+
+func (s *store) GetRankingReferencesByUploadID(ctx context.Context, uploadID int, limit, offset int) (references []shared.RankingReferences, err error) {
```
Comment (efritz) [inline enterprise/internal/codeintel/uploads/internal/lsifstore/lsifstore_ranking.go:286]: This doesn't appear used?

File: enterprise/internal/codeintel/uploads/internal/lsifstore/lsifstore_ranking.go [hunk 21]
```diff
@@ -0,0 +1,375 @@
...
+        jsonb_build_object(cr.path, SUM(count)) AS row
+    FROM combined_ranks cr
+    GROUP BY cr.repository_id, cr.path
+) temp
+GROUP BY temp.repository_id
+ON CONFLICT (repository_id, precision) DO UPDATE SET
+    payload = EXCLUDED.payload
+`
+
+func (s *store) GetRankingDefinitionsBySymbolNames(ctx context.Context, symbolNames []string) (definitions []shared.RankingDefintions, err error) {
```
Comment (efritz) [inline enterprise/internal/codeintel/uploads/internal/lsifstore/lsifstore_ranking.go:242]: This doesn't appear used?

File: enterprise/internal/codeintel/uploads/internal/background/job_graph_exporter.go (hunk 14 above)
Comment (Numbers88s) [reply enterprise/internal/codeintel/uploads/internal/background/job_graph_exporter.go:42]: very :) Updated.

File: lib/codeintel/precise/types.go (hunk 10 above)
Comment (Numbers88s) [reply lib/codeintel/precise/types.go:263]: removed.

File: enterprise/internal/codeintel/uploads/ranking.go (hunk 3 above)
Comment (Numbers88s) [reply enterprise/internal/codeintel/uploads/ranking.go]: This whole thing is removed now. Just answering for posterity.

File: enterprise/internal/codeintel/uploads/internal/lsifstore/lsifstore_ranking.go [hunk 22]
```diff
@@ -0,0 +1,368 @@
...
+const getDocumentsByUploadIDQuery = `
+SELECT
+	sid.document_path,
+	sd.raw_scip_payload
+FROM codeintel_scip_document_lookup sid
+JOIN codeintel_scip_documents sd ON sd.id = sid.document_id
+WHERE sid.upload_id = %s
+ORDER BY sid.document_path
+`
+const batchNumber = 10000
```
Comment (Numbers88s) [inline enterprise/internal/codeintel/uploads/internal/lsifstore/lsifstore_ranking.go:73]: environment variable?

File: enterprise/internal/codeintel/uploads/internal/lsifstore/lsifstore.go [hunk 23]
```diff
@@ -33,8 +35,13 @@ type LsifStore interface {
 	ReconcileCandidates(ctx context.Context, batchSize int) ([]int, error)
 	DeleteUnreferencedDocuments(ctx context.Context, batchSize int, maxAge time.Duration, now time.Time) (count int, err error)
 
-	// Stream
+	// Ranking
 	ScanDocuments(ctx context.Context, id int, f func(path string, document *scip.Document) error) (err error)
```
[the remark this answers is not in the document]
Comment (Numbers88s) [reply enterprise/internal/codeintel/uploads/internal/lsifstore/lsifstore.go:39]: Just a heads up, I rewrote it cause I found the callback pattern a bit confusing and hard to follow

[the remark this answers is not in the document]
Comment (efritz) [reply enterprise/internal/codeintel/uploads/internal/lsifstore/lsifstore.go:39]: We can do a dead code pass later

File: enterprise/internal/codeintel/uploads/ranking.go (hunk 5 above)
Comment (Numbers88s) [reply enterprise/internal/codeintel/uploads/ranking.go]: removed.

File: enterprise/internal/codeintel/uploads/internal/lsifstore/lsifstore_ranking.go (hunk 21 above)
Comment (Numbers88s) [reply enterprise/internal/codeintel/uploads/internal/lsifstore/lsifstore_ranking.go]: removed.

File: enterprise/internal/codeintel/uploads/shared/types.go [hunk 24]
```diff
@@ -199,3 +199,15 @@ type UploadLog struct {
...
+type RankingDefintions struct {
+	UploadID     int
+	SymbolName   string
+	Repository   string
+	DocumentPath string
+}
+
+type RankingReferences struct {
+	UploadID   int
+	SymbolName []string
```
Comment (efritz) [inline enterprise/internal/codeintel/uploads/shared/types.go:212]: Should be `SymbolNames`?

File: enterprise/internal/codeintel/uploads/internal/lsifstore/lsifstore_ranking.go [hunk 25]
```diff
@@ -0,0 +1,330 @@
...
+		tx.Handle(),
+		"codeintel_ranking_references",
+		batch.MaxNumPostgresParameters,
+		[]string{"upload_id", "symbol_names", "graph_key"},
+		inserter,
+	); err != nil {
+		return err
+	}
+
+	fmt.Println("finish inserting all ref batch")
```
Comment (efritz) [inline enterprise/internal/codeintel/uploads/internal/lsifstore/lsifstore_ranking.go:327]: A few prints in this file.

File: enterprise/internal/codeintel/uploads/internal/lsifstore/lsifstore_ranking.go [hunk 26]
```diff
@@ -0,0 +1,330 @@
...
+    WHERE graph_key = %s::text AND NOT processed
+    ORDER BY graph_key, repository, id
+    LIMIT %s
+    FOR UPDATE SKIP LOCKED
+),
+processed AS (
+    UPDATE codeintel_ranking_path_counts_inputs
+    SET processed = true
+    WHERE id IN (SELECT ir.id FROM input_ranks ir)
+	RETURNING 1
```
Comment (efritz) [inline enterprise/internal/codeintel/uploads/internal/lsifstore/lsifstore_ranking.go:237]:
```suggestion
    RETURNING 1
```

Comment (efritz) [reply enterprise/internal/codeintel/uploads/internal/lsifstore/lsifstore_ranking.go:237]: just nits: some strange indentation on L264 as well

File: enterprise/internal/codeintel/uploads/internal/lsifstore/lsifstore_ranking.go [hunk 27]
```diff
@@ -0,0 +1,330 @@
...
+SELECT
+	repository,
+	document_path,
+	COUNT(*),
+	graph_key
+FROM definitions
+GROUP BY repository, document_path, graph_key
+`
+
+func (s *store) InsertPathRanks(ctx context.Context, graphKey string, batchSize int) (numPathRanksInserted float64, numInputsProcessed float64, err error) {
```
Comment (efritz) [inline enterprise/internal/codeintel/uploads/internal/lsifstore/lsifstore_ranking.go:197]: Can we get a unit test here similar to `TestBulkSetAndMergeDocumentRanks`?

Comment (efritz) [reply enterprise/internal/codeintel/uploads/internal/lsifstore/lsifstore_ranking.go:197]: (Will probably test `InsertPathCountInputs` above as part of the setup).

File: enterprise/internal/codeintel/uploads/internal/lsifstore/lsifstore_ranking.go [hunk 28]
```diff
@@ -0,0 +1,330 @@
...
+			END
+	RETURNING 1
+)
+
+SELECT
+	(SELECT COUNT(*) FROM processed) AS num_processed,
+	(SELECT COUNT(*) FROM inserted) AS num_inserted
+`
+
+func (s *store) InsertReferencesForRanking(ctx context.Context, rankingGraphKey string, references shared.RankingReferences) (err error) {
```
Comment (efritz) [inline enterprise/internal/codeintel/uploads/internal/lsifstore/lsifstore_ranking.go:279]: Can we get a simple unit test on each top-level method to confirm the SQL is well-formed (at the very least)?

File: enterprise/internal/codeintel/uploads/internal/lsifstore/lsifstore_ranking.go [hunk 29]
```diff
@@ -0,0 +1,330 @@
...
+
+const insertPathCountInputsQuery = `
+WITH refs AS (
+	SELECT
+		id,
+		unnest(symbol_names) AS symbol_names
+	FROM codeintel_ranking_references rr
+	WHERE rr.graph_key = %s AND NOT rr.processed
+	ORDER BY rr.symbol_names, rr.id
+	LIMIT %s
```
Comment (efritz) [inline enterprise/internal/codeintel/uploads/internal/lsifstore/lsifstore_ranking.go:172]: You'll need to do a `FOR UPDATE` on this CTE

**Review by efritz — COMMENTED** [review]
Everything looks good so far; will do another pass Monday. Thoughts for future me:

- Ensure we change the ranking API to what search team expects
- Ensure we have the same mechanism to remove data from refs/defs as we do in GCS
- Set up grafana monitoring for progress
- Ensure we have an envvar kill switch
- Check indexes against queries in PR

File: migrations/frontend/1676319972_add_codeintel_rank_defintions_and_references/up.sql [hunk 30]
```diff
@@ -0,0 +1,22 @@
+CREATE TABLE codeintel_ranking_definitions (
```
Comment (efritz) [inline migrations/frontend/1676319972_add_codeintel_rank_defintions_and_references/up.sql:1]: `IF NOT EXISTS` here is the reason for (at least some of) the CI errors.

File: enterprise/internal/codeintel/uploads/config.go [hunk 31]
```diff
@@ -118,5 +118,5 @@ var ConfigExportInst = &exportConfig{}
 
 func (c *exportConfig) Load() {
 	c.RankingInterval = c.GetInterval("CODEINTEL_UPLOADS_RANKING_INTERVAL", "1s", "How frequently to serialize a batch of the code intel graph for ranking.")
-	c.NumRankingRoutines = c.GetInt("CODEINTEL_UPLOADS_RANKING_NUM_ROUTINES", "4", "The number of concurrent ranking graph serializer routines to run per worker instance.")
```
Comment (efritz) [inline enterprise/internal/codeintel/uploads/config.go:121]: This should be cool to pump back to 4, right?

File: enterprise/internal/codeintel/uploads/internal/lsifstore/lsifstore_ranking.go (hunk 22 above)
Comment (efritz) [reply enterprise/internal/codeintel/uploads/internal/lsifstore/lsifstore_ranking.go:73]: That or defined in `config.go` next to the service and passed as an arg.

File: migrations/frontend/1676399623_add_codeintel_rank_path_counts_inputs/up.sql [hunk 32]
```diff
@@ -0,0 +1,10 @@
+CREATE TABLE IF NOT EXISTS codeintel_ranking_path_counts_inputs (
+    id bigserial NOT NULL,
```
Comment (efritz) [inline migrations/frontend/1676399623_add_codeintel_rank_path_counts_inputs/up.sql:2]: you'll also need to mark this as a primary key explicitly (`BIGSERIAL PRIMARY KEY`). Same with other tables with id fields.

File: enterprise/internal/codeintel/uploads/internal/lsifstore/lsifstore_ranking.go (hunk 22 above)
Comment (Numbers88s) [reply enterprise/internal/codeintel/uploads/internal/lsifstore/lsifstore_ranking.go:73]: `config.go` it is!

File: enterprise/internal/codeintel/uploads/shared/types.go (hunk 24 above)
Comment (Numbers88s) [reply enterprise/internal/codeintel/uploads/shared/types.go:212]: yup.

File: enterprise/internal/codeintel/uploads/internal/lsifstore/lsifstore_ranking.go (hunk 26 above)
Comment (Numbers88s) [reply enterprise/internal/codeintel/uploads/internal/lsifstore/lsifstore_ranking.go:237]: Fixed. Pushing soon.

File: enterprise/internal/codeintel/uploads/internal/lsifstore/lsifstore_ranking.go (hunk 29 above)
Comment (Numbers88s) [reply enterprise/internal/codeintel/uploads/internal/lsifstore/lsifstore_ranking.go:172]: Added.

File: enterprise/internal/codeintel/uploads/config.go (hunk 31 above)
Comment (Numbers88s) [reply enterprise/internal/codeintel/uploads/config.go:121]: yeah, it should be

File: migrations/frontend/1676399623_add_codeintel_rank_path_counts_inputs/up.sql (hunk 32 above)
Comment (Numbers88s) [reply migrations/frontend/1676399623_add_codeintel_rank_path_counts_inputs/up.sql:2]: Totally, done.
